Task 11. Graph query language

SurfaceYellowDuck opened the pull request:
Add ANTLR grammar of graph query language, also add CI for buil parser by given grammar, and add python functions, that represent the query in the form of a tree, counts node number in a tree and reconstruct a strign by constructed tree.

SurfaceYellowDuck committed:
feat: add parser for gql

SurfaceYellowDuck committed:
feat: add CI steps for build parser

SurfaceYellowDuck committed:
fix: rename antlr to antlr4

SurfaceYellowDuck committed:
fix: fix dependencies

SurfaceYellowDuck committed:
fix: add newline in sh

SurfaceYellowDuck committed:
fix: fix path to GLparser.g4

SurfaceYellowDuck committed:
fix: fix grammar file

KubEF commented on project/hw11/parser.py:
```diff
@@ -0,0 +1,43 @@
+    return tree, True
+
+
+def nodes_count(tree: ParserRuleContext) -> int:
+    if tree is None:
+        return 0
+    count = 0
+    for i in range(tree.getChildCount()):
+        child = tree.getChild(i)
+        count += nodes_count(child)
```
Давайте через `visitor`/`listener`, а не через рекурсию

KubEF commented on project/hw11/parser.py:
```diff
@@ -0,0 +1,43 @@
+    return count
+
+
+def tree_to_program(tree: ParserRuleContext) -> str:
+    if tree is None:
+        return ""
+
+    reconstructed_string = ""
+    for child in tree.getChildren():
+        if isinstance(child, TerminalNode):
+            if child.getText() != "<EOF>":
+                reconstructed_string += child.getText() + " "
+            else:
+                reconstructed_string += "\n"
+        else:
+            reconstructed_string += tree_to_program(child)
+    return reconstructed_string
```
Давайте через `visitor`/`listener`, а не через рекурсию

SurfaceYellowDuck committed:
fix: removed recursion
- changed recursive methods of building string to visitor/listener.
- samily with counting nodes of tree

SurfaceYellowDuck committed:
fix:removed  trailing whitespace
